### ISSUE TITLE
Drawing controls: next steps

### ISSUE THREAD

**Issue by mstenta** [farmOS/farmOS-map#40]
Follow up to: https://github.com/farmOS/farmOS-map/issues/18

Let's use this issue to discuss next steps for the farmOS-map drawing controls, and break them out into separate issues as needed.

Here are some we already talked about in #18:

- Snap interaction (#24)
- Use SVG icons instead of Unicode characters for buttons.
- Add visual separation between the draw buttons and the select/delete buttons.

**Comment by mstenta** [comment]
From https://github.com/farmOS/farmOS-map/issues/18#issuecomment-533644501:

> right now selecting, modifying, and moving are all enabled at the same time, which works, but sometimes you end up performing a move when you want to modify and vice versa, especially when using your fingers. So one thing we could consider is splitting those out into separate buttons that are available when a feature is selected (like the delete button).

**Comment by mstenta** [comment]
Feedback from Will in Slack:

> - I'd increase the size of the hitbox for point addition/modification, right now it's so small that you often end up moving the polygon when you intend to edit or add a point
> - I'd let hitting the Escape key exit drawing mode, so you don't have to finish drawing a polygon if you mess up
> - The "X" button communicates "Exit drawing mode" to me. But it's actually a Delete button, so I'd make it a trash can
> - The Line button looks like a minus sign, when I look at it I would assume it deletes a point in a polygon. I'd switch the icon to more like a "/" with dots at endpoints
> - of course it'd be awesome to not have to switch modes explicitly to modify a features, though I know this would be more work to implement

**Comment by mstenta** [comment]
Created a new issue for SVG icons specifically: https://github.com/farmOS/farmOS-map/issues/41

**Comment by jgaehring** [comment]
Playing around with modifying features makes me think that we'll definitely want the ability to remove vertices, too. I wound up with a really funky geometry because when I thought I was tapping on a vertex to move it, I ended up creating a new vertex instead. Wound up with a really jagged shape in the end, with a lot of unnecessary points. This could probably be helped by Will's suggestion, too:

> I'd increase the size of the hitbox for point addition/modification

Also,
> Add visual separation between the draw buttons and the select/delete buttons.

I'm not sure we need this, and we could also end up crunched for space this way on mobile. Unless we did something like a divider, or a really subtle difference in spacing, like an additional 2px of padding between the drawing and the modifying buttons.

I think overall, we could stand to do an overhaul of the styling to make it mobile-first.

**Comment by mstenta** [comment]
> I'd increase the size of the hitbox for point addition/modification

FYI here is the OL API documentation for the `Modify` interaction:

https://openlayers.org/en/latest/apidoc/module-ol_interaction_Modify-Modify.html

It has an option `pixelTolerance` option which might be what we want to dial in.  It defaults to `10`.

**Comment by jgaehring** [comment]
> It has an option pixelTolerance option which might be what we want to dial in. It defaults to 10.

Ah yes, I thought I remembered something like that, looks nice n easy!

**Comment by mstenta** [comment]
Also: we should add the "Circle" draw button that we currently have in farmOS. This is useful for drawing things like pivot irrigated fields.

**Comment by mstenta** [comment]
Similar to the ESC key idea for cancelling a drawing interaction, I would like to be able to click the drawing button a second time for the same purpose. So, when you click "Polygon" for example, and you begin drawing, you can click "Polygon" again to cancel the turn off the drawing interaction and cancel any unfinished shape you had begun drawing.

**Comment by mstenta** [comment]
Will (in person) also just suggested changing the default button colors so they match the farmOS color scheme a little closer.

**Comment by mstenta** [comment]
> we should add the "Circle" draw button that we currently have in farmOS. This is useful for drawing things like pivot irrigated fields.

Done: https://github.com/farmOS/farmOS-map/commit/af3b51f25e76bae18102083cc7e6e30cc9584834

> Add visual separation between the draw buttons and the select/delete buttons.

Done: https://github.com/farmOS/farmOS-map/commit/88fae7db7068d0e9d7bc7b187603c825f8f05efa

**Comment by jgaehring** [comment]
> Add visual separation between the draw buttons and the select/delete buttons.

Do you have a screenshot for what this looks like?

I'm wondering if 1em is too much...

**Comment by mstenta** [comment]
Here it is with `0.75em`:

![Screenshot from 2019-10-30 11-03-43](https://user-images.githubusercontent.com/95381/67870284-faf03d00-fb04-11e9-97cf-3593abae3fb5.png)

Maybe that's a better middle ground?

**Comment by mstenta** [comment]
Ah I'm just re-reading your comments above...

> I'm not sure we need this, and we could also end up crunched for space this way on mobile.

Yea, it does get crunched on mobile, you're right. But not just because of the spacer... the drawing buttons in general are poorly positioned on small screens:

![Screenshot from 2019-10-30 11-14-34](https://user-images.githubusercontent.com/95381/67871292-80282180-fb06-11e9-87a6-59991b6df8ca.png)

**Comment by jgaehring** [comment]
I was thinking more like `.25em`. I think we're just going for a "visual cue" that distinguishes a difference in purpose between the buttons. In other words, don't think there's an actual need to have them far apart, neither for ergonomic nor layout reasons. So I think any perceptible difference in the spacing will suffice, and the human eye is pretty good at detecting pretty subtle differences like that. In fact, I'd think even just twice the space as there is between same-group buttons would be enough.

Another thing we could do is add a divider, like Google Docs does in its toolbar:

![image](https://user-images.githubusercontent.com/5556545/67871343-933af180-fb06-11e9-9cdc-6d4dbcdfb6f6.png)

**Comment by mstenta** [comment]
What do you think of this?

![Screenshot from 2019-10-30 11-19-02](https://user-images.githubusercontent.com/95381/67871670-1eb48280-fb07-11e9-877d-8ff22656d099.png)

- Moved them to the side (below zoom and geocoder), vertically oriented.
- `0.5em` separation between draw and actions

**Comment by mstenta** [comment]
Also, for reference, this is what the controls look like in farmOS currently (provided by the Drupal OpenLayers module... not using the new farmOS-map library):

![Screenshot from 2019-10-30 11-25-13](https://user-images.githubusercontent.com/95381/67872160-f5e0bd00-fb07-11e9-8caa-91f88c06fe75.png)

**Comment by jgaehring** [comment]
![image](https://user-images.githubusercontent.com/5556545/67872604-9800a500-fb08-11e9-9d4c-7a94aaa0ddf4.png)

In my opinion...

A == 2(B)
C == A

**Comment by jgaehring** [comment]
Again, we just need a visual cue, and I think even a few pixels difference will suffice.

**Comment by jgaehring** [comment]
> Here it is with C = 4px, which I think makes it roughly equal to the default border around all buttons (A):

That looks a lot better to me.

**Comment by mstenta** [comment]
What do you think about this?

> right now selecting, modifying, and moving are all enabled at the same time, which works, but sometimes you end up performing a move when you want to modify and vice versa, especially when using your fingers. So one thing we could consider is splitting those out into separate buttons that are available when a feature is selected (like the delete button).

Doing this would fix the issue Will pointed out (without needing to mess with default hit boxes):

> I'd increase the size of the hitbox for point addition/modification, right now it's so small that you often end up moving the polygon when you intend to edit or add a point

I can mock it up real quick and see how they look as separate buttons. Standby for screenshot.

**Comment by jgaehring** [comment]
> So one thing we could consider is splitting those out into separate buttons that are available when a feature is selected (like the delete button).

Yes, I think this would help a lot! Although I don't think we need 3 separate buttons for Select, Modify, and Move. Instead I think the best UX would be to have 2 buttons: a button for Select & Modify, and a button for Select & Move. I think that eliminates the conflicts w/o necessitating more taps by the user.

**Comment by jgaehring** [comment]
Cool, looks great!

**Comment by mstenta** [comment]
Couple of general improvements committed:

- Split "Select/modify/move" button into "Modify" and "Move" buttons (with implicit select): https://github.com/mstenta/farmOS-map/commit/667b8c6c7b14645212ba792f7432be9bc5870187
- If an active button is clicked, deactivate it: https://github.com/mstenta/farmOS-map/commit/dcae00f89414ee67fd5f09cf4cc22524f427e4e1
- Deselect all features whenever a button is clicked: https://github.com/mstenta/farmOS-map/commit/6104af68bb56e6bfcd767831d060c5054a056b8f
